lasagne/layers/pool.py

```diff
@@ -755,3 +756,161 @@ def get_output_shape_for(self, input_shape):
 
     def get_output_for(self, input, **kwargs):
         return self.pool_function(input.flatten(3), axis=2)
+
+
+def pooling_fixed_output_size(inputs, output_size, pooling_op=T.max):
+    """
+    Performs a pooling operation that results in a fixed output size.
+    Used by SpatialPyramidPoolingLayer. Refer to appendix A in [1]
+
+    Parameters
+    ----------
+    inputs : a tensor with 4 dimensions (N x C x H x W)
+    output_size: integer
+        The output size of the pooling operation
+    pooling_op: a theano op
+        The pooling operation (default: max pooling)
+
+    Returns a tensor with 3 dimensions (N x C x output_size*output_size)
+
+    References
+    ----------
+    .. [1] He, Kaiming et al (2015):
+           Spatial Pyramid Pooling in Deep Convolutional Networks
+           for Visual Recognition.
+           http://arxiv.org/pdf/1406.4729.pdf.
+    """
+    h, w = inputs.shape[2:]
+
+    result = []
+    n = float(output_size)
+
+    for row in range(output_size):
+        for col in range(output_size):
+            start_h = T.floor(row / n * h).astype('int32')
+            end_h = T.ceil((row + 1) / n * h).astype('int32')
+            start_w = T.floor(col / n * w).astype('int32')
+            end_w = T.ceil((col + 1) / n * w).astype('int32')
+
+            pooling_region = inputs[:, :, start_h:end_h, start_w:end_w]
+            this_result = pooling_op(pooling_region, axis=(2, 3))
+            result.append(this_result.dimshuffle(0, 1, 'x'))
+    return T.concatenate(result, axis=2)
+
+
+class SpatialPyramidPoolingLayer(Layer):
+    """
+    Spatial Pyramid Pooling Layer
+
+    Performs spatial pyramid pooling (SPP) over the input.
+    It will turn a 2D input of arbitrary size into an output of fixed
+    dimension.
+    Hence, the convolutional part of a DNN can be connected to a dense part
+    with a fixed number of nodes even if the dimensions of the
+    input image are unknown.
+
+    The pooling is performed over :math:`l` pooling levels.
+    Each pooling level :math:`i` will create :math:`M_i` output features.
+    :math:`M_i` is given by :math:`n_i * n_i`,
+    with :math:`n_i` as the number of pooling operation per dimension in
+    level :math:`i`, and we use a list of the :math:`n_i`'s as a
+    parameter for SPP-Layer.
+    The length of this list is the level of the spatial pyramid.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or tuple
+        The layer feeding into this layer, or the expected input shape.
+
+    pool_dims : list of integers
+        The list of :math:`n_i`'s that define the output dimension of each
+        pooling level :math:`i`. The length of pool_dims is the level of
+        the spatial pyramid.
+
+    mode : string
+        Pooling mode, one of 'max', 'average_inc_pad', 'average_exc_pad' or
+        'kaiming'. Defaults to 'max'.
```

Comment:
Sorry, didn't have a look at the interface before. If `mode` is already used to distinguish between max- and mean-pooling, then `'kaiming'` vs. `'naive'` (or whatever we want to call the current implementation) should become a different keyword argument. Maybe `implementation`. `implementation='kaiming`' should also allow mean-pooling (but it doesn't need to distinguish between `'average_inc_pad'` and `'average_exc_pad'`).